sdk/core/Azure.Core/src/Internal/ModelWriter.SequenceBuilder.cs

```diff
@@ -0,0 +1,166 @@
+    {
+        private sealed class SequenceBuilder : IBufferWriter<byte>, IDisposable
+        {
+            private struct Buffer
+            {
+                public byte[] Array;
+                public int Written;
+            }
+
+            private volatile Buffer[] _buffers; // this is an array so items can be accessed by ref
+            private volatile int _count;
+            private int _segmentSize;
+
+            /// <summary>
+            /// Initializes a new instance of <see cref="SequenceBuilder"/>.
+            /// </summary>
+            /// <param name="segmentSize">The size of each buffer segment.</param>
+            public SequenceBuilder(int segmentSize = 4096)
+            {
+                _segmentSize = segmentSize;
+                _buffers = Array.Empty<Buffer>();
+            }
+
+            /// <summary>
+            /// Notifies the <see cref="SequenceBuilder"/> that bytes bytes were written to the output <see cref="Span{T}"/> or <see cref="Memory{T}"/>.
+            /// You must request a new buffer after calling <see cref="Advance(int)"/> to continue writing more data; you cannot write to a previously acquired buffer.
+            /// </summary>
+            /// <param name="bytesWritten">The number of bytes written to the <see cref="Span{T}"/> or <see cref="Memory{T}"/>.</param>
+            /// <exception cref="ArgumentOutOfRangeException"></exception>
+            public void Advance(int bytesWritten)
+            {
+                ref Buffer last = ref _buffers[_count - 1];
+                last.Written += bytesWritten;
+                if (last.Written > last.Array.Length)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(bytesWritten));
+                }
+            }
+
+            /// <summary>
+            /// Returns a <see cref="Memory{T}"/> to write to that is at least the requested size, as specified by the <paramref name="sizeHint"/> parameter.
+            /// </summary>
+            /// <param name="sizeHint">The minimum length of the returned <see cref="Memory{T}"/>. If less than 256, a buffer of size 256 will be returned.</param>
+            /// <returns>A memory buffer of at least <paramref name="sizeHint"/> bytes. If <paramref name="sizeHint"/> is less than 256, a buffer of size 256 will be returned.</returns>
+            public Memory<byte> GetMemory(int sizeHint = 0)
+            {
+                if (sizeHint < 256)
+                {
+                    sizeHint = 256;
+                }
+
+                int sizeToRent = sizeHint > _segmentSize ? sizeHint : _segmentSize;
+
+                if (_buffers.Length == 0)
+                {
+                    ExpandBuffers(sizeToRent);
+                }
+
+                ref Buffer last = ref _buffers[_count - 1];
+                Memory<byte> free = last.Array.AsMemory(last.Written);
+                if (free.Length >= sizeHint)
+                {
+                    return free;
```

Comment:
What would happen if two threads get access to the same `Buffer.Array` concurrently?

Reply:
This is an private implementation class where we control threading via the public surface on ModelWriter.  Until this becomes internal or public we should not need to worry about this.  If you think somehow a customer could do this via the public surface can you help me understand how?